kythe/go/extractors/validation/validation.go

```diff
@@ -0,0 +1,225 @@
+	fromKzip, err := h.filenamesFromKzip(h.Kzip)
+	if err != nil {
+		return nil, fmt.Errorf("reading from kzip %s: %v", h.Kzip, err)
+	}
+	fromRepo, err := h.filenamesFromPath(h.Repo)
+	if err != nil {
+		return nil, fmt.Errorf("reading from repo %s: %v", h.Repo, err)
+	}
+	missingFromKzip := map[string]bool{}
+	for f := range fromRepo {
+		if _, ok := fromKzip[f]; !ok {
+			missingFromKzip[f] = true
+		}
+	}
+
+	// Try to calculate the most missing subdirectories.
+	topMissing := Coverage{}
+	// Only try to go so many paths deep (src/java/com/domain/sub).
+	// TODO(danielmoy): be more intelligent about this.
+	for i := 1; i <= topMissingPathsDepth; i++ {
+		topMissing = getTopMissing(missingFromKzip, topMissingPathsLimit, i)
+		if len(topMissing.Paths) >= topMissingPathsLimit {
+			break
+		}
+	}
+
+	if h.MissingOutput != nil {
+		err = printMissing(missingFromKzip, *h.MissingOutput)
+	}
+
+	return &Result{
+		FilesInKzip: len(fromKzip),
+		FilesInRepo: len(fromRepo),
+		Missing:     len(missingFromKzip),
+		TopMissing:  topMissing,
+	}, err
+}
+
+func (h Harness) filenamesFromKzip(path string) (map[string]bool, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("opening kzip %s: %v", path, err)
+	}
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("getting kzip size %s: %v", path, err)
+	}
+	r, err := kzip.NewReader(f, fi.Size())
+	if err != nil {
+		return nil, fmt.Errorf("opening kzip reader %s: %v", path, err)
+	}
+
+	ret := map[string]bool{}
+
+	if err := r.Scan(func(cu *kzip.Unit) error {
+		if cu.Proto != nil {
+			for _, f := range cu.Proto.SourceFile {
+				ret[f] = true
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("scanning kzip %s: %v", path, err)
+	}
+
+	return ret, nil
+}
+
+func (h Harness) filenamesFromPath(path string) (map[string]bool, error) {
+	ret := map[string]bool{}
+	err := filepath.Walk(path, func(p string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		ext := filepath.Ext(p)
+		if len(ext) > 0 && ext[0] == '.' {
+			ext = ext[1:]
+		}
+		if _, ok := h.Langs[ext]; ok {
+			// We are only interested in the repo-relative path.
+			rp, err := filepath.Rel(path, p)
+			if err != nil {
+				return err
+			}
+			ret[rp] = true
+		}
+		return nil
+	})
+	return ret, err
+}
+
+// Returns the top N most common Mth-level subdirectories.
+func getTopMissing(missing map[string]bool, num, level int) Coverage {
+	counts := map[string]int{}
+	for f := range missing {
+		dirs := strings.Split(f, string(os.PathSeparator))
+		if len(dirs) > level-1 {
+			counts[strings.Join(dirs[:level-1], string(os.PathSeparator))]++
+		} else {
+			counts[f]++
+		}
+	}
+
+	top := Coverage{}
+	for f, c := range counts {
+		top.Paths = append(top.Paths, Dir{f, c})
+	}
+	sort.Sort(sort.Reverse(&top))
+	if len(top.Paths) >= num {
+		top.Paths = top.Paths[:num]
+	}
+	return top
+}
+
+func printMissing(filenames map[string]bool, output string) (err error) {
```

Comment:
rename to outputMissing?

Reply:
It will be easier to test if, instead of creating a file here you take an io.Writer and let the caller decide.

Then a test can use a bytes.Buffer, or os.Stdout, or whatever.

Reply:
Yup yup, same deal where I used to be printing it to console but that ended up being terrible, and I forgot to rename stuff after fixing.

Reply:
Noted, will consider that for future change.